app/controllers/oauth/oauth_controller.rb

```diff
@@ -54,12 +93,68 @@ def token
 
     private
 
+    def validate_user_decision(decision)
+      raise ArgumentError, "Missing decision parameter" if decision.blank?
+      raise ArgumentError, "Invalid decision" unless %w[allow deny].include?(decision)
+    end
+
+    def handle_access_denied(state_data)
+      Redis.current.del("oauth_state:#{params[:state]}")
+
+      error_redirect = build_redirect_url(state_data["redirect_uri"], {
+        error: "access_denied",
+        error_description: "User denied authorization"
```

Comment:
Unless you want to include another parameter to include a reasoning behind an good/bad `callback` argument I wouldn't include this, a boolean `true`/`false` decision on whether or not your credential provider is telling you to auth the user isn't the same as a reason for that decision.